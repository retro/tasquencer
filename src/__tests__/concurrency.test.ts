import { Effect } from 'effect';
import { it } from 'vitest';

import { Builder, IdGenerator, Service } from '../index.js';
import { getEnabledTaskNames, makeIdGenerator } from './shared.js';

const workflowDefinition = Builder.workflow()
  .withName('activities')
  .startCondition('start')
  .task('t1', (t) =>
    t()
      .withWorkItem((w) =>
        w<number>().onStart(({ startWorkItem, getWorkItem }) =>
          Effect.gen(function* ($) {
            const { enqueueCompleteWorkItem } = yield* $(startWorkItem());
            const workItem = yield* $(getWorkItem());
            yield* $(Effect.sleep(`${workItem.payload * 100} millis`));
            yield* $(enqueueCompleteWorkItem());
          })
        )
      )
      .onEnable(({ enableTask }) =>
        Effect.gen(function* ($) {
          const { enqueueStartTask } = yield* $(enableTask());
          yield* $(enqueueStartTask());
        })
      )
      .onStart(({ startTask }) =>
        Effect.gen(function* ($) {
          const { initializeWorkItem, enqueueStartWorkItem } = yield* $(
            startTask()
          );
          for (const i of [1, 2, 3]) {
            const { id } = yield* $(initializeWorkItem(i));
            yield* $(enqueueStartWorkItem(id));
          }
        })
      )
  )
  .endCondition('end')
  .connectCondition('start', (to) => to.task('t1'))
  .connectTask('t1', (to) => to.condition('end'));

it('will start work items concurrently and emit state change on each work item start', async ({
  expect,
}) => {
  const program = Effect.gen(function* ($) {
    const idGenerator = makeIdGenerator();
    const logs: unknown[] = [];

    const service = yield* $(
      workflowDefinition.build(),
      Effect.flatMap((workflow) => Service.initialize(workflow)),
      Effect.provideService(IdGenerator, idGenerator)
    );

    service.onStateChange((changes) => {
      return Effect.succeed(logs.push(changes));
    });

    yield* $(service.start());
    const state = yield* $(service.getState());
    expect(state).toMatchSnapshot();
    expect(logs).toMatchSnapshot();
    expect(getEnabledTaskNames(state)).toEqual(new Set());
    expect(state.workflows[0]?.state).toEqual('completed');
  });

  await Effect.runPromise(program);
});
