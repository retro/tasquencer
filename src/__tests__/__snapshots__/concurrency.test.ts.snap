// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`will start work items concurrently and emit state change on each work item start 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": 1,
      "state": "completed",
      "taskGeneration": 1,
      "taskName": "t1",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "id": "workItem-2",
      "payload": 2,
      "state": "completed",
      "taskGeneration": 1,
      "taskName": "t1",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "id": "workItem-3",
      "payload": 3,
      "state": "completed",
      "taskGeneration": 1,
      "taskName": "t1",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "completed",
    },
  ],
}
`;

exports[`will start work items concurrently and emit state change on each work item start 2`] = `
[
  [
    {
      "change": {
        "type": "WORKFLOW_INITIALIZED",
        "workflow": {
          "context": undefined,
          "id": "workflow-1",
          "name": "activities",
          "parent": null,
          "state": "initialized",
        },
      },
      "getState": [Function],
    },
    {
      "change": {
        "task": {
          "generation": 0,
          "name": "t1",
          "state": "disabled",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
        "type": "TASK_INITIALIZED",
      },
      "getState": [Function],
    },
    {
      "change": {
        "condition": {
          "marking": 0,
          "name": "start",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
        "type": "CONDITION_INITIALIZED",
      },
      "getState": [Function],
    },
    {
      "change": {
        "condition": {
          "marking": 0,
          "name": "end",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
        "type": "CONDITION_INITIALIZED",
      },
      "getState": [Function],
    },
  ],
  [
    {
      "change": {
        "type": "WORKFLOW_STATE_UPDATED",
        "workflow": {
          "context": undefined,
          "id": "workflow-1",
          "name": "activities",
          "parent": null,
          "state": "started",
        },
      },
      "getState": [Function],
    },
    {
      "change": {
        "condition": {
          "marking": 1,
          "name": "start",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
        "type": "CONDITION_MARKING_UPDATED",
      },
      "getState": [Function],
    },
    {
      "change": {
        "task": {
          "generation": 0,
          "name": "t1",
          "state": "enabled",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
        "type": "TASK_STATE_UPDATED",
      },
      "getState": [Function],
    },
  ],
  [
    {
      "change": {
        "task": {
          "generation": 1,
          "name": "t1",
          "state": "started",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
        "type": "TASK_STATE_UPDATED",
      },
      "getState": [Function],
    },
    {
      "change": {
        "condition": {
          "marking": 0,
          "name": "start",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
        "type": "CONDITION_MARKING_UPDATED",
      },
      "getState": [Function],
    },
  ],
  [
    {
      "change": {
        "type": "WORK_ITEM_INITIALIZED",
        "workItem": {
          "id": "workItem-1",
          "payload": 1,
          "state": "initialized",
          "taskGeneration": 1,
          "taskName": "t1",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
      },
      "getState": [Function],
    },
    {
      "change": {
        "type": "WORK_ITEM_INITIALIZED",
        "workItem": {
          "id": "workItem-2",
          "payload": 2,
          "state": "initialized",
          "taskGeneration": 1,
          "taskName": "t1",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
      },
      "getState": [Function],
    },
    {
      "change": {
        "type": "WORK_ITEM_INITIALIZED",
        "workItem": {
          "id": "workItem-3",
          "payload": 3,
          "state": "initialized",
          "taskGeneration": 1,
          "taskName": "t1",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
      },
      "getState": [Function],
    },
  ],
  [
    {
      "change": {
        "type": "WORK_ITEM_STATE_UPDATED",
        "workItem": {
          "id": "workItem-1",
          "payload": 1,
          "state": "started",
          "taskGeneration": 1,
          "taskName": "t1",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
      },
      "getState": [Function],
    },
  ],
  [
    {
      "change": {
        "type": "WORK_ITEM_STATE_UPDATED",
        "workItem": {
          "id": "workItem-2",
          "payload": 2,
          "state": "started",
          "taskGeneration": 1,
          "taskName": "t1",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
      },
      "getState": [Function],
    },
  ],
  [
    {
      "change": {
        "type": "WORK_ITEM_STATE_UPDATED",
        "workItem": {
          "id": "workItem-3",
          "payload": 3,
          "state": "started",
          "taskGeneration": 1,
          "taskName": "t1",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
      },
      "getState": [Function],
    },
  ],
  [
    {
      "change": {
        "type": "WORK_ITEM_STATE_UPDATED",
        "workItem": {
          "id": "workItem-1",
          "payload": 1,
          "state": "completed",
          "taskGeneration": 1,
          "taskName": "t1",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
      },
      "getState": [Function],
    },
  ],
  [
    {
      "change": {
        "type": "WORK_ITEM_STATE_UPDATED",
        "workItem": {
          "id": "workItem-2",
          "payload": 2,
          "state": "completed",
          "taskGeneration": 1,
          "taskName": "t1",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
      },
      "getState": [Function],
    },
  ],
  [
    {
      "change": {
        "type": "WORK_ITEM_STATE_UPDATED",
        "workItem": {
          "id": "workItem-3",
          "payload": 3,
          "state": "completed",
          "taskGeneration": 1,
          "taskName": "t1",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
      },
      "getState": [Function],
    },
    {
      "change": {
        "task": {
          "generation": 1,
          "name": "t1",
          "state": "completed",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
        "type": "TASK_STATE_UPDATED",
      },
      "getState": [Function],
    },
    {
      "change": {
        "condition": {
          "marking": 1,
          "name": "end",
          "workflowId": "workflow-1",
          "workflowName": "activities",
        },
        "type": "CONDITION_MARKING_UPDATED",
      },
      "getState": [Function],
    },
    {
      "change": {
        "type": "WORKFLOW_STATE_UPDATED",
        "workflow": {
          "context": undefined,
          "id": "workflow-1",
          "name": "activities",
          "parent": null,
          "state": "completed",
        },
      },
      "getState": [Function],
    },
  ],
]
`;
